cli/pack/app_dir.go

```diff
@@ -81,6 +112,83 @@ func initAppDir(appDirPath string, ctx *context.Ctx) error {
 	return nil
 }
 
+func copyPathFromCache(cachedPath string, destPath string) error {
+	log.Infof("Using cached path %s", filepath.Base(destPath))
+	err := copy.Copy(cachedPath, destPath)
+
+	if err != nil {
+		return fmt.Errorf("Failed to copy path %s from cache to project directory: %s", destPath, err)
+	}
+
+	return nil
+}
+
+func getProjectCachePaths(ctx *context.Ctx) (*cacheProjectPaths, error) {
+	if ctx.Pack.NoCache {
+		return nil, nil
+	}
+
+	rockspecPath, err := common.FindRockspec(ctx.Project.Path)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to find rockspec: %s", err)
+	} else if rockspecPath == "" {
+		return nil, fmt.Errorf("Application directory should contain rockspec")
+	}
+
+	projectPathHash := common.StringSHA1Hex(ctx.Project.Path)[:10]
+	rockspecHash, err := common.FileSHA1Hex(rockspecPath)
+	if err != nil {
+		return nil, err
+	}
+
+	rockspecHash = rockspecHash[:10]
+
+	return &cacheProjectPaths{
+		projectPath: filepath.Join(ctx.Cli.CacheDir, projectPathHash, rocksDirPath),
+		cachePath:   filepath.Join(ctx.Cli.CacheDir, projectPathHash, rocksDirPath, rockspecHash)}, nil
+}
+
+func rotateCacheDirs(ctx *context.Ctx) error {
+	dir, err := ioutil.ReadDir(ctx.Cli.CacheDir)
+	if err != nil {
+		return err
+	}
+
+	if len(dir) > maxCachedProjects {
+		// Sort project cache paths by change time
+		sort.Slice(dir, func(i, j int) bool {
+			return dir[i].ModTime().Before(dir[j].ModTime())
+		})
+
+		for i := 0; i < len(dir)-maxCachedProjects; i++ {
+			projectCachePath := filepath.Join(ctx.Cli.CacheDir, dir[i].Name())
+			log.Debugf("Removing project cache directory: %s", projectCachePath)
+			os.RemoveAll(projectCachePath)
+		}
+	}
+
+	return nil
+}
+
+func updateCache(ctx *context.Ctx, cachePaths *cacheProjectPaths, path string) error {
+	if ctx.Pack.NoCache {
+		return nil
+	}
+
+	if _, err := os.Stat(cachePaths.projectPath); err == nil {
+		if err := common.ClearDir(cachePaths.projectPath); err != nil {
```

Comment:
Why do you clear all project caches when only `cachePath` should be updated?

Reply:
BTW, I'm not sure that you need to store `.projectPath`. It's used only when paths to cached rocks are computed. Then, you work with each cache separately. Maybe `cachePaths` can be a dictionary?

```go
cachePaths = map[string]string{
    ".rocks": "<path-computed-using-projectPath-and-key>"
}
```